File: examples/example_ca.py
#!/usr/bin/python2

from math import *
from gi.repository import GObject
import matplotlib.pyplot as plt
from gi.repository import NumCosmo as Nc
from gi.repository import NumCosmoMath as Ncm

#
#  Initializing the library objects, this must be called before 
#  any other library function.
#
Ncm.cfg_init ()

#
#  New homogeneous and isotropic cosmological model NcHICosmoDEXcdm 
#
cosmo = Nc.HICosmo.new_from_name (Nc.HICosmo, "NcHICosmoDEXcdm")

#
#  New cosmological distance objects optimizied to perform calculations
#  up to redshift 2.0.
#
dist = Nc.Distance.new (2.0)

#
# New windown function 'NcWindowTophat'
#
wp =  Nc.Window.new_from_name ("NcWindowTophat")

#
# New transfer function 'NcTransferFuncEH' using the Einsenstein, Hu
# fitting formula.
#
tf = Nc.TransferFunc.new_from_name ("NcTransferFuncEH")

#
# New matter variance object using FFT method for internal calculations and
# the window and transfer functions defined above.
#
vp = Nc.MatterVar.new (Nc.MatterVarStrategy.FFT, wp, tf)

#
# New growth function
#
gf = Nc.GrowthFunc.new ()

#
# New multiplicity function 'NcMultiplicityFuncTinkerMean'
#
mulf = Nc.MultiplicityFunc.new_from_name ("NcMultiplicityFuncTinkerMean")

#
# New mass function object using the objects defined above.
#
mf = Nc.MassFunction.new (dist, vp, gf, mulf)

#
#  Setting values for the cosmological model, those not set stay in the
#  default values. Remember to use the _orig_ version to set the original
#  parameters when a reparametrization is used.
#
cosmo.props.H0      = 70.0
cosmo.props.Omegab  = 0.05
cosmo.props.Omegac  = 0.25
cosmo.props.Omegax  = 0.70
cosmo.props.Tgamma0 = 2.72
cosmo.props.ns      = 1.0
cosmo.props.sigma8  = 0.9
cosmo.props.w       = -1.0

#
#  Printing the parameters used.
#
print "# Model parameters: ", 
cosmo.params_log_all ()

#
# Number of points to build the plots
#
np = 2000
divfac = 1.0 / (np - 1.0)

#
#  Calculating growth and its derivative in the [0, 2] redshift
#  range.
#
za = []
Da = []
dDa = []

gf.prepare (cosmo)

for i in range (0, np):
  z = 2.0 * divfac * i
  D = gf.eval (cosmo, z)
  dD = gf.eval_deriv (cosmo, z)
  za.append (z)
  Da.append (D)
  dDa.append (dD)
  print z, D

#
#  Ploting growth function.
#

plt.title ("Growth Function")
plt.plot (za, Da, 'r', label="D")
plt.plot (za, dDa, 'b--', label="dD/dz")
plt.xlabel('$z$')
plt.legend(loc=2)

plt.savefig ("growth_func.png")
plt.clf ()

#
# Calculating the transfer function and the matter power spectrum in the
# kh (in unities of h/Mpc) interval [1e-3, 1e3]
#

kha = []
Ta = []
Pma = []

tf.prepare (cosmo)

for i in range (0, np):
  lnkh = log (1e-4) +  log (1e7) * divfac * i
  kh = exp (lnkh)
  T = tf.eval (cosmo, kh)
  Pm = 1.0e3 / 7.0 * tf.matter_powerspectrum (cosmo, kh)
  kha.append (kh)
  Ta.append (T)
  Pma.append (Pm)

#
#  Ploting transfer and matter power spectrum
#

plt.title ("Transfer Function and Matter Power Spectrum")
plt.xscale('log')
plt.plot (kha, Ta, 'r', label="T(kh)")
plt.plot (kha, Pma, 'b--', label="P_m(kh)")
plt.xlabel('$k$')
plt.legend(loc=1)

plt.savefig ("transfer_func.png")
plt.clf ()

#
# Calculating the variance filtered with the tophat windown function using
# scales R in the interval [5, 50] at redshift 0.3.
# First calculates the growth function at z = 0.3 and then the spectrum
# amplitude from the sigma8 parameter.
#

vp.prepare (cosmo)

Dz = gf.eval (cosmo, 0.3)
A  = vp.sigma8_sqrtvar0 (cosmo)
prefact = A * A * Dz * Dz

Ra = []
sigma2a = []
dlnsigma2a = []

for i in range (0, np):
  lnR = log (5.0) +  log (10.0) * divfac * i
  R = exp (lnR)
  sigma2 = prefact * vp.var0 (cosmo, lnR)
  dlnsigma2 = vp.dlnvar0_dlnR (cosmo, lnR) 
  Ra.append (R)
  sigma2a.append (sigma2)
  dlnsigma2a.append (dlnsigma2)

#
#  Ploting filtered matter variance
#

plt.title ("Variance and Variance Derivative")
plt.plot (Ra, sigma2a, 'r', label='$\sigma^2(\ln(R))$')
plt.plot (Ra, dlnsigma2a, 'b--', label='$d\ln(\sigma^2)/d\ln(R)$')
plt.xlabel('$R$')
plt.legend(loc=1)

plt.savefig ("matter_var.png")
plt.clf ()

#
# Calculating the mass function integrated in the mass interval [1e14, 1e16]
# for the redhshifts in the interval [0, 2.0] and area 200 squared degree.
#

mf.set_area_sd (200.0)
mf.set_eval_limits (cosmo, log (1e14), log(1e16), 0.0, 2.0)

dndza = []

for i in range (0, np):
  dndz = mf.dn_dz (cosmo, log(1e14), log(1e16), za[i], True)
  dndza.append (dndz)

#
#  Ploting the mass function
#

plt.title ("Mass Function")
plt.plot (za, dndza, 'r', label='$dn/dz$')
plt.xlabel('$z$')
plt.legend(loc=1)

plt.savefig ("mass_function.png")
plt.clf ()

File: examples/example_ca.c
#include <glib.h>
#include <numcosmo/numcosmo.h>

gint
main (gint argc, gchar *argv[])
{
  NcHIPrim  *prim;
  NcHIReion *reion;
  NcHICosmo *cosmo;
  NcDistance *dist;
  NcTransferFunc *tf;
  NcMultiplicityFunc *mulf;
  NcHaloMassFunction *mf;
  NcPowspecML *psml;
  NcmPowspecFilter *psf;
  guint np = 50;
  gint i;

  /**************************************************************************** 
   * Initializing the library objects, this must be called before 
   * any other library function.
   ****************************************************************************/  
  ncm_cfg_init ();
  
  /**************************************************************************** 
   * New homogeneous and isotropic cosmological model NcHICosmoDEXcdm.
   ****************************************************************************/  
  cosmo = nc_hicosmo_new_from_name (NC_TYPE_HICOSMO, "NcHICosmoDEXcdm");

  /**************************************************************************** 
   * New homogeneous and isotropic reionization object.
   ****************************************************************************/  
  reion = NC_HIREION (nc_hireion_camb_new ());

  /**************************************************************************** 
   * New homogeneous and isotropic primordial object.
   ****************************************************************************/  
  prim  = NC_HIPRIM (nc_hiprim_power_law_new ());

  /**************************************************************************** 
   * Adding the submodels to the main cosmology model.
   ****************************************************************************/  
  ncm_model_add_submodel (NCM_MODEL (cosmo), NCM_MODEL (reion));
  ncm_model_add_submodel (NCM_MODEL (cosmo), NCM_MODEL (prim));

  /**************************************************************************** 
   * New cosmological distance objects optimizied to perform calculations
   * up to redshift 2.0.
   ****************************************************************************/  
  dist = nc_distance_new (2.0);
 
  /**************************************************************************** 
   * New transfer function 'NcTransferFuncEH' using the Einsenstein and Hu
   * fitting formula.
   ****************************************************************************/  
  tf = nc_transfer_func_new_from_name ("NcTransferFuncEH");

  /**************************************************************************** 
   * New linear matter power spectrum object based of the EH transfer function.
   ****************************************************************************/  
  psml = NC_POWSPEC_ML (nc_powspec_ml_transfer_new (tf));
  ncm_powspec_require_kmin (NCM_POWSPEC (psml), 1.0e-3);
  ncm_powspec_require_kmax (NCM_POWSPEC (psml), 1.0e3);

  /**************************************************************************** 
   * Apply a tophat filter to the psml object, set best output interval.
   ****************************************************************************/     
  psf = ncm_powspec_filter_new (NCM_POWSPEC (psml), NCM_POWSPEC_FILTER_TYPE_TOPHAT);
  ncm_powspec_filter_set_best_lnr0 (psf);

  /**************************************************************************** 
   * New multiplicity function 'NcMultiplicityFuncTinkerMean'
   ****************************************************************************/  
  mulf = nc_multiplicity_func_tinker_new_full (NC_MULTIPLICITY_FUNC_MASS_DEF_MEAN, 200.0);
  
  /**************************************************************************** 
   * New mass function object using the objects defined above.
   ****************************************************************************/  
  mf = nc_halo_mass_function_new (dist, psf, mulf);

  /**************************************************************************** 
   * Setting values for the cosmological model, those not set stay in the
   * default values. Remeber to use the _orig_ version to set the original
   * parameters in case when a reparametrization is used.
   ****************************************************************************/ 
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_H0,        70.0);
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_OMEGA_C,    0.25);
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_OMEGA_X,    0.7);
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_T_GAMMA0,   1.0);
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_OMEGA_B,    0.05);
  ncm_model_orig_param_set (NCM_MODEL (cosmo), NC_HICOSMO_DE_XCDM_W,    -1.1);

  /**************************************************************************** 
   * Printing the parameters used.
   ****************************************************************************/
  printf ("# Model parameters:\n#"); 
  ncm_model_params_log_all (NCM_MODEL (cosmo));

  /**************************************************************************** 
   * Printing the growth function and its derivative with respect to z 
   * up to redshift 1.
   ****************************************************************************/ 
  ncm_powspec_prepare (NCM_POWSPEC (psml), NCM_MODEL (cosmo));

  {
    NcGrowthFunc *gf = nc_powspec_ml_transfer_peek_gf (NC_POWSPEC_ML_TRANSFER (psml));
    for (i = 0; i < np; i++)
    {
      gdouble z = 1.0 / (np - 1.0) * i;
      gdouble gfz  = nc_growth_func_eval (gf, cosmo, z);
      gdouble dgfz = nc_growth_func_eval_deriv (gf, cosmo, z);
      printf ("% 10.8f % 21.15g % 21.15g\n", z, gfz, dgfz);
    }
    printf ("\n\n");
  }

  /**************************************************************************** 
   * Printing the matter power spectrum in the k (in unities of Mpc^-1) 
   * interval [1e-3, 1e3] 
   ****************************************************************************/

  for (i = 0; i < np; i++)
  {
    gdouble lnk = log (1e-3) + log (1e6) / (np - 1.0) * i;
    gdouble Pmk = ncm_powspec_eval (NCM_POWSPEC (psml), NCM_MODEL (cosmo), 0.0, exp (lnk));
    printf ("% 10.8f % 21.15g\n", exp (lnk), Pmk);
  }
  printf ("\n\n");

  /**************************************************************************** 
   * Printing the variance filtered with the tophat windown function using
   * scales R in the interval [5, 50] at redshift 0.3.
   * First calculates the growth function at z = 0.3 and then the spectrum
   * amplitude from the sigma8 parameter.
   ****************************************************************************/
  ncm_powspec_filter_prepare (psf, NCM_MODEL (cosmo));
  {
    for (i = 0; i < np; i++)
    {
      gdouble lnR = log (5.0) + log (10.0) / (np - 1.0) * i;
      gdouble sigma2       = ncm_powspec_filter_eval_var_lnr (psf, 0.0, lnR);
      gdouble dsigma2_dlnR = ncm_powspec_filter_eval_dvar_dlnr (psf, 0.0, lnR);
      printf ("% 10.8f % 21.15g % 21.15g\n", exp (lnR), sigma2, dsigma2_dlnR);
    }
    printf ("\n\n");
  }

  /**************************************************************************** 
   * Printing the mass function integrated in the mass interval [1e14, 1e16]
   * for the redhshifts in the interval [0, 2.0] and area 200 squared degree.
   ****************************************************************************/
  nc_halo_mass_function_set_area_sd (mf, 200.0);
  nc_halo_mass_function_set_eval_limits (mf, cosmo, log (1e14), log (1e16), 0.0, 2.0);
  nc_halo_mass_function_prepare (mf, cosmo);

  for (i = 0; i < np; i++)
  {
    gdouble z = 2.0 / (np - 1.0) * i;
    gdouble dndz = nc_halo_mass_function_dn_dz (mf, cosmo, log (1.0e14), log (1.0e16), z, FALSE);
    printf ("% 10.8f % 21.15g\n", z, dndz);
  }
  printf ("\n\n");


  /**************************************************************************** 
   * Freeing objects.
   ****************************************************************************/ 
  nc_distance_free (dist);
  ncm_model_free (NCM_MODEL (cosmo));
  nc_hireion_free (reion);
  nc_transfer_func_free (tf);
  nc_powspec_ml_free (psml);
  ncm_powspec_filter_free (psf);
  nc_multiplicity_func_free (mulf);
  nc_halo_mass_function_free (mf);

  return 0;
}
